prefetcher/EIP/EIP.cc

```diff
@@ -0,0 +1,871 @@
+
+bool l1i_ongoing_request(uint64_t line_addr) { return l1i_find_timing_mshr_entry(line_addr) < L1I_TIMING_MSHR_SIZE; }
+
+bool l1i_ongoing_accessed_request(uint64_t line_addr)
+{
+  uint32_t index = l1i_find_timing_mshr_entry(line_addr);
+  if (index == L1I_TIMING_MSHR_SIZE)
+    return false;
+  return l1i_timing_mshr_table[l1i_cpu_id][index].accessed;
+}
+
+uint64_t l1i_get_latency_timing_mshr(uint64_t line_addr)
+{
+  uint32_t index = l1i_find_timing_mshr_entry(line_addr);
+  if (index == L1I_TIMING_MSHR_SIZE)
+    return 0;
+  if (!l1i_timing_mshr_table[l1i_cpu_id][index].accessed)
+    return 0;
+  return l1i_get_latency(current_core_cycle[l1i_cpu_id], l1i_timing_mshr_table[l1i_cpu_id][index].timestamp);
+}
+
+// RECORD ENTANGLED TABLE
+
+uint32_t L1I_ENTANGLED_FORMATS[L1I_ENTANGLED_MAX_FORMATS] = {58, 28, 18, 13, 10, 8, 6};
+#define L1I_ENTANGLED_NUM_FORMATS 6
+
+uint32_t l1i_get_format_entangled(uint64_t line_addr, uint64_t entangled_addr)
+{
+  for (uint32_t i = L1I_ENTANGLED_NUM_FORMATS; i != 0; i--) {
+    if ((line_addr >> L1I_ENTANGLED_FORMATS[i - 1]) == (entangled_addr >> L1I_ENTANGLED_FORMATS[i - 1])) {
+      return i;
+    }
+  }
+  assert(false);
+}
+
+uint64_t l1i_extend_format_entangled(uint64_t line_addr, uint64_t entangled_addr, uint32_t format)
+{
+  return ((line_addr >> L1I_ENTANGLED_FORMATS[format - 1]) << L1I_ENTANGLED_FORMATS[format - 1])
+         | (entangled_addr & (((uint64_t)1 << L1I_ENTANGLED_FORMATS[format - 1]) - 1));
+}
+
+uint64_t l1i_compress_format_entangled(uint64_t entangled_addr, uint32_t format)
+{
+  return entangled_addr & (((uint64_t)1 << L1I_ENTANGLED_FORMATS[format - 1]) - 1);
+}
+
+#define L1I_ENTANGLED_TABLE_INDEX_BITS 8
+#define L1I_ENTANGLED_TABLE_SETS (1 << L1I_ENTANGLED_TABLE_INDEX_BITS)
+#define L1I_ENTANGLED_TABLE_WAYS 34
+#define L1I_MAX_ENTANGLED_PER_LINE L1I_ENTANGLED_NUM_FORMATS
+#define L1I_TAG_BITS (42 - L1I_ENTANGLED_TABLE_INDEX_BITS)
+#define L1I_TAG_MASK (((uint64_t)1 << L1I_TAG_BITS) - 1)
+#define L1I_CONFIDENCE_COUNTER_BITS 2
+#define L1I_CONFIDENCE_COUNTER_MAX_VALUE ((1 << L1I_CONFIDENCE_COUNTER_BITS) - 1)
+#define L1I_CONFIDENCE_COUNTER_THRESHOLD 1
+
+#define L1I_TRIES_AVAIL_ENTANGLED 6
+#define L1I_TRIES_AVAIL_ENTANGLED_NOT_PRESENT 2
+
+typedef struct __l1i_entangled_entry {
+  uint64_t tag;                                        // L1I_TAG_BITS bits
+  uint32_t format;                                     // 3 bits
+  uint64_t entangled_addr[L1I_MAX_ENTANGLED_PER_LINE]; // keep just diff
+  uint32_t entangled_conf[L1I_MAX_ENTANGLED_PER_LINE]; // L1I_CONFIDENCE_COUNTER_BITS bits
+  uint32_t bb_size;                                    // L1I_MERGE_BBSIZE_BITS bits
+} l1i_entangled_entry;
+
+l1i_entangled_entry l1i_entangled_table[NUM_CPUS][L1I_ENTANGLED_TABLE_SETS][L1I_ENTANGLED_TABLE_WAYS];
+uint32_t l1i_entangled_fifo[NUM_CPUS][L1I_ENTANGLED_TABLE_SETS]; // log2(L1I_ENTANGLED_TABLE_WAYS) * L1I_ENTANGLED_TABLE_SETS bits
```

Comment:
If we're using FIFO replacement, I think it would be better for this to be:
```suggestion
std::array<std::array<std::deque<l1i_entangled_entry>, L1I_ENTAGLED_TABLE_SETS>, NUM_CPUS> l1i_entangled_table;
```